lib/spell-check-manager.coffee

```diff
@@ -0,0 +1,384 @@
+    # then locales any changed locales or known words checker.
+    if changed
+      @emitSettingsChanged()
+
+  emitSettingsChanged: ->
+    if @isTask
+      emit("spell-check:settings-changed")
+
+  addCheckerPath: (checkerPath) ->
+    checker = require checkerPath
+    @addPluginChecker checker
+
+  addPluginChecker: (checker) ->
+    # Add the spell checker to the list.
+    @addSpellChecker checker
+
+    # We only emit a settings change for plugins since the core checkers are
+    # handled in a different manner.
+    @emitSettingsChanged()
+
+  addSpellChecker: (checker) ->
+    @checkers.push checker
+
+  removeSpellChecker: (spellChecker) ->
+    @checkers = @checkers.filter (plugin) -> plugin isnt spellChecker
+
+  check: (args, text) ->
+    # Make sure our deferred initialization is done.
+    @init()
+
+    # We need a couple packages.
+    multirange = require 'multi-integer-range'
+
+    # For every registered spellchecker, we need to find out the ranges in the
+    # text that the checker confirms are correct or indicates is a misspelling.
+    # We keep these as separate lists since the different checkers may indicate
+    # the same range for either and we need to be able to remove confirmed words
+    # from the misspelled ones.
+    correct = new multirange.MultiRange([])
+    incorrects = []
+
+    for checker in @checkers
+      # We only care if this plugin contributes to checking spelling.
+      if not checker.isEnabled() or not checker.providesSpelling(args)
+        continue
+
+      # Get the results which includes positive (correct) and negative (incorrect)
+      # ranges. If we have an incorrect range but no correct, everything not
+      # in incorrect is considered correct.
+      results = checker.check(args, text)
+
+      if results.invertIncorrectAsCorrect and results.incorrect
+        # We need to add the opposite of the incorrect as correct elements in
+        # the list. We do this by creating a subtraction.
+        invertedCorrect = new multirange.MultiRange([[0, text.length]])
+        removeRange = new multirange.MultiRange([])
+        for range in results.incorrect
+          removeRange.appendRange(range.start, range.end)
+        invertedCorrect.subtract(removeRange)
+
+        # Everything in `invertedCorrect` is correct, so add it directly to
+        # the list.
+        correct.append invertedCorrect
+      else if results.correct
+        for range in results.correct
+          correct.appendRange(range.start, range.end)
+
+      if results.incorrect
+        newIncorrect = new multirange.MultiRange([])
+        incorrects.push(newIncorrect)
+
+        for range in results.incorrect
+          newIncorrect.appendRange(range.start, range.end)
+
+    # If we don't have any incorrect spellings, then there is nothing to worry
+    # about, so just return and stop processing.
+    misspellings = []
+
+    if incorrects.length is 0
+      return {id: args.id, misspellings}
```

Comment:
What do you think about returning an empty array here without using the `misspellings` variable? Like:


```coffee
return {id: args.id, misspellings, []}
```

Then we can defer the variable declaration to later when we actually need it.